README.md

```diff
@@ -1,47 +1,121 @@
+Pulling or building the containers for the first time may take several minutes to complete, and the FHIR servers take a minute or two to start up in every case. Once everything is up and running, you should be able to access the various services and interfaces through `localhost`!
 
-Now `curl http://localhost:9000/api/fhir/Patient?` should return sample data!
+### Available services
 
-By default, you'll have servers running at ports:
+This list shows all services and interfaces included in the reference stack, along with default ports which can be adjusted by editing `docker-compose.override.yml`.
 
- * `9000`: SMART EHR server
- * `9006`: SMART EHR server (Pointing at HAPI FHIR server version STU3)
- * `9001`: Demo app server
- * `9002`: For debugging only, the internal HAPI FHIR server
- * `9007`: For debugging only, the internal HAPI FHIR server (STU3)
- * `9003`: For testing only, the test suite client
+| Name                 | Description                   | Port   |
+|----------------------|-------------------------------|--------|
+| `api`                | HAPI-FHIR server (DSTU2)      | `9002` |
+| `api-stu3`           | HAPI-FHIR server (STU3)       | `9005` |
+| `smart`              | SMART layer (DSTU2)           | `9000` |
+| `smart-stu3`         | SMART layer (STU3)            | `9006` |
+| `app`                | Research application frontend | `9001` |
+| `research-app-api`   | Research application backend  | `9005` |
+| `token-introspector` | Token introspection service   | `9004` |
+| `tests`              | Sync for Science test suite   | `9003` |
 
+## Components
 
-## Update submodules to latest on GH branches
- * `git submodule update --remote`
+### Vendor
 
+#### HAPI-FHIR servers (`api`/`api-stu3`)
 
----
+[HAPI-FHIR](http://hapifhir.io/) is an implementation of the FHIR specification written in Java. We include both STU2 and DSTU3 FHIR servers, which provide the base FHIR functionality for the stack.
 
-# Deploy in prod
+#### SMART on FHIR (`smart`/`smart-stu3`)
 
-For example, see [./deploys/demo.syncfor.science](./deploys/demo.syncfor.science)
+We provide a layer that sits on top of the HAPI-FHIR servers and provides some of the functionality prescribed by [SMART on FHIR](http://docs.smarthealthit.org/). This layer acts as an EHR vendor's patient-facing interface, where a participant would be directed to in order to log into the vendor's site and consent to providing a research application with access to their data.
+
+Configuration options in `docker-compose.override.yml`:
+ * `BASE_URL`: This should be set to the URL that the service is being accessed by, which is either `http://localhost:9000`/`http://localhost:9006` when a human is using the research application interfaces, or `http://smart:5000`/`http://smart-stu3:5000` when the test suite is testing the authentication process, because the test suite accesses the SMART interfaces directly through Docker networking.
```

Comment:
This is a bit confusing -- probably because the underlying implementation is confusing :-)

Is this saying that `BASE_URL` for some services should be `localhost...` and for other services should be `smart...`? Or that depending on what a user is doing, they might need to change `BASE_URL` across the board?

(I think the right fix for the test suite may to have an `INTERNAL_BASE_URL` and an `EXTERNAL_BASE_URL`, or somesuch, to allow for full configuration.)

Reply:
I struggled with a better way to set this up but didn't quite get there. The issue is that some of the SMART URLs come from the the `BASE_URL` config variable via the conformance statement (snippet from auth-proxy project):

        authorize_url = url_for('oauth.cb_oauth_authorize', _external=True)
        manage_url = url_for('main.apps', _external=True)

        base_url = current_app.config['BASE_URL']
        if base_url:
            authorize_url =  base_url + '/oauth/authorize'
            manage_url =  base_url + '/apps'

This was added because `url_for` will build the URL based on the request context, and since the request for the conformance statement is made though the internal Docker network from the `research-app-api` container, the URL is built starting with `http://smart:5000/`. A user who is interacting with the app through their browser will then be redirected to this host which is unavailable outside of Docker. Setting `BASE_URL` to `http://localhost:9000/` and using this to rewrite those URLs allows the process to work correctly for the human user.

However in the case of the test suite, the "user" interacting with the site is actually the test suite code running from inside Docker. In this case, the browser needs to be redirected to the `smart` host rather than `localhost`. We could introduce two `BASE_URL`s but there still would need some way to identify if we need the internal or the external URL in the conformance statement. I thought about adding some logic through `research-app-api` and possibly `app` to look at the requested host to identify this, but I didn't really like the idea of including code in those projects that is so specific to the way our test-suite has to interact with the services due to the Docker setup.

Reply:
Thanks for the thoughtful assessment here! I'm also not enamored of making deep code changes to support our local dev/testing workflow, but a small amount to this kind of instrumentation can go a long way. I'd be  happy to defer until later (with a new issue in the backlog) or address here, at your preference.